File: rust/src/leetcode/Q417_pacific_atlantic_water_flow.rs
/*
 * @lc app=leetcode id=417 lang=rust
 *
 * [417] Pacific Atlantic Water Flow
 */

// @lc code=start
impl Solution {
    fn flood(heights: &Vec<Vec<i32>>, map: &mut Vec<Vec<bool>>, row: i32, col: i32) {
        static DIRECTIONS: [(i32, i32); 4] = [(-1, 0), (1, 0), (0, -1), (0, 1)];

        map[row as usize][col as usize] = true;
        let height: i32 = heights[row as usize][col as usize];

        for (dx, dy) in DIRECTIONS {
            let new_row: i32 = row + dx;
            let new_col: i32 = col + dy;

            if new_row >= 0
                && new_col >= 0
                && new_row < heights.len() as i32
                && new_col < heights[0].len() as i32
                && !map[new_row as usize][new_col as usize]
                && height <= heights[new_row as usize][new_col as usize]
            {
                Self::flood(heights, map, new_row, new_col);
            }
        }
    }

    pub fn pacific_atlantic(heights: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        let rows: usize = heights.len();
        let cols: usize = heights[0].len();
        let mut pacific_map: Vec<Vec<bool>> = vec![vec![false; cols]; rows];
        let mut atlantic_map: Vec<Vec<bool>> = vec![vec![false; cols]; rows];

        for row in 0..rows {
            Self::flood(&heights, &mut pacific_map, row as i32, 0);
            Self::flood(&heights, &mut atlantic_map, row as i32, cols as i32 - 1);
        }

        for col in 0..cols {
            Self::flood(&heights, &mut pacific_map, 0, col as i32);
            Self::flood(&heights, &mut atlantic_map, rows as i32 - 1, col as i32);
        }

        let mut result: Vec<Vec<i32>> = Vec::new();

        for row in 0..rows {
            for col in 0..cols {
                if pacific_map[row][col] && atlantic_map[row][col] {
                    result.push(vec![row as i32, col as i32]);
                }
            }
        }

        result
    }
}
// @lc code=end

File: python/Q417_pacific_atlantic_water_flow.py
#
# @lc app=leetcode id=417 lang=python3
#
# [417] Pacific Atlantic Water Flow
#


# @lc code=start
class Solution:
    def dfs(self, row: int, col: int, water_map: list[list[bool]]) -> None:
        for i, j in self.directions:
            if (
                0 <= row + i < self.ROWS
                and 0 <= col + j < self.COLS
                and not water_map[row + i][col + j]
                and self.heights[row + i][col + j] >= self.heights[row][col]
            ):
                water_map[row + i][col + j] = True
                self.dfs(row + i, col + j, water_map)

    def check_pacific(self) -> None:
        visiting: set[tuple[int, int]] = set()

        for i in range(self.ROWS):
            self.pacific[i][0] = True
            visiting.add((i, 0))

        for j in range(self.COLS):
            self.pacific[0][j] = True
            visiting.add((0, j))

        for i, j in visiting:
            self.dfs(i, j, self.pacific)

    def check_atlantic(self) -> None:
        visiting: set[tuple[int, int]] = set()

        for i in range(self.ROWS):
            self.atlantic[i][self.COLS - 1] = True
            visiting.add((i, self.COLS - 1))

        for j in range(self.COLS):
            self.atlantic[self.ROWS - 1][j] = True
            visiting.add((self.ROWS - 1, j))

        for i, j in visiting:
            self.dfs(i, j, self.atlantic)

    def pacificAtlantic(self, heights: list[list[int]]) -> list[list[int]]:
        self.ROWS: int = len(heights)
        self.COLS: int = len(heights[0])
        self.heights: list[list[int]] = heights

        self.directions: list[tuple[int, int]] = [(1, 0), (-1, 0), (0, 1), (0, -1)]

        self.pacific: list[list[bool]] = [[False] * self.COLS for _ in range(self.ROWS)]
        self.atlantic: list[list[bool]] = [
            [False] * self.COLS for _ in range(self.ROWS)
        ]

        self.check_pacific()
        self.check_atlantic()

        result: list[list[int]] = []

        for i in range(self.ROWS):
            for j in range(self.COLS):
                if self.pacific[i][j] and self.atlantic[i][j]:
                    result.append([i, j])

        return result


# @lc code=end
